lib/rspec-console/config_cache.rb

```diff
@@ -1,82 +1,84 @@
+          add(
+            :main,
+            shared_examples.keys.first,
+            &shared_examples.values.first
+          )
         end
+      else
+        ::RSpec.world.
+          shared_example_groups.merge!(recorded_registry || {}) rescue nil
       end
-      ::RSpec.world.shared_example_groups.merge!(self.shared_examples_groups || {}) rescue nil
-
     else
       # record
-      real_config = ::RSpec.configuration
-      self.recorded_config = []
-      self.proxy = Proxy.new(self.recorded_config, real_config)
+      original_config = ::RSpec.configuration
+      self.proxy = RSpecConsole::Proxy.new(original_config)
+
       ::RSpec.configuration = self.proxy
-      yield
-      ::RSpec.configuration = real_config
-      self.shared_examples_groups = ::RSpec.world.shared_example_groups.dup rescue nil
 
-      # rspec-rails/lib/rspec/rails/view_rendering.rb add methods on the
-      # configuration singleton. Need advice to copy them without going down
-      # the road with object2module.
-    end
+      yield # spec helper is called during this yield, see #reset
 
-    # Well, instead of copying them, we redirect them to the configuration
-    # proxy. Looks like it good enough.
-    proxy = self.proxy
-    ::RSpec.configuration.singleton_class.send(:define_method, :method_missing) do |method, *args, &block|
-      proxy.send(method, *args, &block)
-    end
+      ::RSpec.configuration = original_config
+
+      if version >= Gem::Version.new('3')
+        self.recorded_registry = ::RSpec.world.
+          shared_example_group_registry.dup rescue nil
+      else
+        self.recorded_registry = ::RSpec.world.
+          shared_example_groups.dup rescue nil
+      end
 
+      # forward to proxy object by delegating to it on any missing method
+      ::RSpec.configuration.singleton_class.send(:define_method, :method_missing) do |method, *args, &block|
+        # note this is not called until runtime when a method is not found on RSpec.configuration
+        self.proxy.send(method, *args, &block)
+      end
+    end
   end
-end
 
-class Proxy < Struct.new(:output, :target)
-  [:include, :extend].each do |method|
-    define_method(method) do |*args|
-      method_missing(method, *args)
+  private
+  def replay_recorded_config
+    Proc.new do |config|
+      self.proxy.output.each do |msg|
+        config.send(msg[:method], *msg[:args], &msg[:block])
+      end
     end
   end
 
-  def method_missing(method, *args, &block)
-    self.output << {:method => method, :args => args, :block => block}
-    self.target.send(method, *args, &block)
+  def have_recording?
+    self.proxy
   end
-end
 
-# For compatibility with Ruby 1.8.x outside of the Rails framework
-if RUBY_VERSION =~ /1.8/
-  unless defined?(Rails)
-    class Object
-      def singleton_class
-        class << self; self; end
-      end
-    end
+  def shared_examples
+    recorded_registry.
+      send(:shared_example_groups)[:main] rescue nil
```

Comment:
it is merely an accessor on the state of the class. what i find about it weird now in retrospect is it doesn't really do what's advertised, because it only checks shared examples for rspec 3, not either version. so i will see about rearranging.